### PR TITLE
Added external cached resources usage, viewer API version update.

### PR DESCRIPTION
**Pull request by kegor**
Replicated changes from: groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC#65.

Related changes in UI: groupdocs-total/GroupDocs.Total-Angular#201.

### CODE REVIEW

File: src/Products/Viewer/Cache/CacheExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+            if (!cache.TryGetValue(key, out entry))
+            {
+                entry = acquire();
+                cache.Set(key, entry);
+            }
+
+            return entry;
+        }
+
+        public delegate TEntry Func<TEntry>();
```
Comment (egorovpavel) [inline src/Products/Viewer/Cache/CacheExtensions.cs:25]: Codacy found an issue: [Add the 'out' keyword to parameter 'TEntry' to make it 'covariant'.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Total-for-NET-MVC/pullRequest?prid=6016413)

**Comment by egorovpavel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/Products/Viewer/Cache/CrossProcessLock.cs  3
- src/Products/Viewer/Cache/FileViewerCache.cs  4
- src/Products/Viewer/Cache/PngViewer.cs  4
- src/Products/Viewer/Cache/HtmlViewer.cs  4
- src/Products/Viewer/Cache/CacheExtensions.cs  2
         

Complexity decreasing per file
==============================
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -4
         

Clones added
============
- src/Products/Viewer/Cache/PngViewer.cs  3
- src/Products/Viewer/Cache/HtmlViewer.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GroupDocs/GroupDocs.Total-for-NET-MVC/pullRequest?prid=6016413&bid=19721162)

File: src/Products/Viewer/Cache/CacheExtensions.cs (hunk 1 above)
Comment (bobkovalex) [inline src/Products/Viewer/Cache/CacheExtensions.cs:25]: Codacy found an issue: [Add the 'out' keyword to parameter 'TEntry' to make it 'covariant'.](https://app.codacy.com/gh/groupdocs-total/GroupDocs.Total-for-NET-MVC/pullRequest?prid=6016412)

**Comment by bobkovalex** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/Products/Viewer/Cache/PngViewer.cs  4
- src/Products/Viewer/Cache/CacheExtensions.cs  2
- src/Products/Viewer/Cache/HtmlViewer.cs  4
- src/Products/Viewer/Cache/FileViewerCache.cs  4
- src/Products/Viewer/Cache/CrossProcessLock.cs  3
         

Complexity decreasing per file
==============================
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -4
         

Clones added
============
- src/Products/Viewer/Cache/PngViewer.cs  3
- src/Products/Viewer/Cache/HtmlViewer.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/groupdocs-total/GroupDocs.Total-for-NET-MVC/pullRequest?prid=6016412&bid=19721161)

**Comment by bobkovalex** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/Products/Viewer/Cache/CrossProcessLock.cs  3
- src/Products/Viewer/Cache/CacheExtensions.cs  2
- src/Products/Viewer/Cache/HtmlViewer.cs  4
- src/Products/Viewer/Cache/FileViewerCache.cs  4
- src/Products/Viewer/Cache/PngViewer.cs  4
         

Complexity decreasing per file
==============================
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -4
         

Clones added
============
- src/Products/Viewer/Cache/HtmlViewer.cs  3
- src/Products/Viewer/Cache/PngViewer.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/groupdocs-total/GroupDocs.Total-for-NET-MVC/pullRequest?prid=6016412&bid=19721161)

**Comment by egorovpavel** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/Products/Viewer/Cache/FileViewerCache.cs  4
- src/Products/Viewer/Cache/CrossProcessLock.cs  3
- src/Products/Viewer/Cache/HtmlViewer.cs  4
- src/Products/Viewer/Cache/PngViewer.cs  4
- src/Products/Viewer/Cache/CacheExtensions.cs  2
         

Complexity decreasing per file
==============================
+ src/Products/Viewer/Controllers/ViewerApiController.cs  -4
         

Clones added
============
- src/Products/Viewer/Cache/HtmlViewer.cs  3
- src/Products/Viewer/Cache/PngViewer.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GroupDocs/GroupDocs.Total-for-NET-MVC/pullRequest?prid=6016413&bid=19721162)

**Pull request merged**